core/trino-spi/src/main/java/io/trino/spi/connector/RowChangeParadigm.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+/**
+ * Different connectors have different ways of representing row updates,
+ * imposed by the underlying storage systems.  The  Trino engine classifies
+ * these different paradigms as elements of this RowChangeParadigm
+ * enumeration, returned by {@link ConnectorMetadata#getRowChangeParadigm}
+ */
+public enum RowChangeParadigm
```

Comment:
If a connector implements copy-on-write (i believe Raptor does that, and Delta Lake too, maybe others), which paradigm should they choose?

cc @losipiuk @alexjo2144

Reply:
If the connector implements row update as a row delete followed by a row insert, `DELETE_ROW_AND_INSERT_ROW` seems appropriate.  My understanding is that Delta Lake implements UPDATE by delaying writing until the first change is encountered, and then copying from the beginning of the file.  Since it doesn't need to separate deletes and inserts resulting from updates, it could use `CHANGE_ONLY_UPDATED_COLUMNS`.

I originally had an enum instance `COPY_ON_WRITE`, but didn't implement the `MergeProcessor` because I wasn't sure what the connector would want.  @electrum suggested it be removed until there is a user.

Reply:
> My understanding is that Delta Lake implements UPDATE by delaying writing until the first change is encountered, and then copying from the beginning of the file.

I am not familiar with what exactly Delta Lake is implemented and I am not fluent in reading Spark Scala code.

What i am trying to say is that there are examples of data sources that have copy-on-write modification styles ad Raptor, Delta Lake and Iceberg v1 are examples of these (to the best of my knowledge).

I don't know yet how a copy-on-write connector is expected to be fitted into the scheme that's provided here. In particular, we still want to avoid copying whole data (and actually: reading the whole data) if only one row in one file is modified; or nothing gets actually modified.

It's probably fine not to have a concrete implementation of a `MergeProcessor` and I actually don't even know if a dedicated `MergeProcessor` would be needed. What's important is to know that the existing design is extensible to accommodate future needs. Maybe all this fits in some of the existing paradigms that are already implemented? Please advice.

Reply:
Copy-on-write is something the connector needs to handle. Since modified rows can arrive at the writer in any order, unlike the `DELETE` and `UPDATE` design that uses `UpdatablePageSource` on the same machine, I believe the only viable strategy is to have the connector buffer all of the deletes into memory or a temporary file. At the end, it can rewrite the file by removing the deleted rows. The inserts can be written into a separate file.

Normal filtering or pushdown should avoid reading unnecessary rows during the deletion operation. However, if there are any actually deleted rows, the rewrite would likely have to read the entire file, unless it can use the deleted row IDs to optimize the read for the rewrite.

Reply:
We could consider implementing `MERGE` for Raptor in order to have an example of copy-on-write, and it would be nice from a developer perspective since you can run it normally in the IDE.

Reply:
Good idea.  I'll add it to my list.